drivers: stepper: api: add stop function

jilaypandya opened the pull request:
add stop function

@jbehrensnx please review this PR :)
#86052

jilaypandya committed:
drivers: stepper: api: add stop function

add stop function
update enable & set_interval specification

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

dipakgmx commented:
A couple of remarks from my end:
- is it really necessary add such complexity into the stop function. A stop is a stop. Am not able to figure out a use case for commanding a stop and then expecting the motor to stop at a later point of time.
- Could we add a callback to the stop function, so the user is informed upon the motor coming to a halt

jilaypandya commented:
> A couple of remarks from my end:
> 
> * is it really necessary add such complexity into the stop function. A stop is a stop. Am not able to figure out a use case for commanding a stop and then expecting the motor to stop at a later point of time.
> * Could we add a callback to the stop function, so the user is informed upon the motor coming to a halt

Probably there could be a separate function for decelerating  and reaching to velocity 0 usteps_per_second with ramp drivers. @jbehrensnx ?

bjarki-andreasen commented on include/zephyr/drivers/stepper.h:
```diff
@@ -95,6 +95,16 @@ enum stepper_event {
+ * @brief Stepper Stop Modes
+ */
+enum stepper_stop_mode {
+	/** If stepper driver supports ramping, start decelerating */
```
For motors which need to "soft stop", meaning there is enough inertia for the motor to stall if "hard stopped", "hard stop" is not actually a valid option. For ramping controllers, the ramp should simply be the most aggressive it can be for the motor to not stall implicitly when stop is called :)

I would remove the enum, and just have `void`. It will make the API more portable since its one less "decision" branch, "decision" in quotes here since its not really an option to stall the motor :)

jbehrensnx replied:
The idea behind soft and hard stop was that the first would result in a controlled stop, while the hard stop would exert the maximal breaking power possible, accepting stalling or another undefined state as long as the motor comes to a stop as fast as possible.

jbehrensnx commented:
> > A couple of remarks from my end:
> > 
> > * is it really necessary add such complexity into the stop function. A stop is a stop. Am not able to figure out a use case for commanding a stop and then expecting the motor to stop at a later point of time.
> > * Could we add a callback to the stop function, so the user is informed upon the motor coming to a halt
> 
> Probably there could be a separate function for decelerating and reaching to velocity 0 usteps_per_second with ramp drivers. @jbehrensnx ?

Decelerating towards 0 usteps/sec and stop could be 2 different functions. It would mean an additional api function however.
I would implicitly use the existing `STEPPER_EVENT_STEPS_COMPLETED` event, but defining a new `STEPPER_EVENT_STOPPED` event would also be an option in my opinion. This would allow for using the existing callback handling for stopping.

jilaypandya committed:
drivers: stepper: gpio: implement stop function

Implement sample stop function and update enable function

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jilaypandya committed:
drivers: stepper: shell: add stop function

Add stop function to stepper shell. align the function order to match
the one in __subsystem stepper_driver_api struct

Signed-off-by: Jilay Pandya <jilay.pandya@outlook.com>

jbehrensnx commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -328,15 +328,23 @@ static int gpio_stepper_enable(const struct device *dev, bool enable)
+			const int err = energize_coils(dev, true);
+
+			if (err != 0) {
+				return -EIO;
```
This causes a system crash. You should simply assign the error value to a return variable instead. This error is repeated multiple times in the driver, but this is not the PR for fixing the existing occurrences.

jbehrensnx commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -346,6 +354,27 @@ static int gpio_stepper_enable(const struct device *dev, bool enable)
+		const int err = energize_coils(dev, true);
+
+		if (err != 0) {
+			return -EIO;
```
Same issue as above

bjarki-andreasen reviewed (dismissed):
Nice :)

jbehrensnx commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -327,20 +327,36 @@ static int gpio_stepper_set_event_callback(const struct device *dev,
+			data->callback(data->dev, STEPPER_EVENT_STOPPED, data->event_cb_user_data);
 		}
 	}
 	return 0;
```
Shouldn't the possible error code be returned? Currently the return value is 0 even if there were errors. I'd argue, that the missing stop event is not enough.

jilaypandya replied:
you are right, my bad, have corrected this in the latest update

fabiobaltieri commented on drivers/stepper/stepper_shell.c:
```diff
@@ -195,6 +198,30 @@ static int cmd_stepper_enable(const struct shell *sh, size_t argc, char **argv)
+static int cmd_stepper_stop(const struct shell *sh, size_t argc, char **argv)
+{
+	const struct device *dev;
+	int err = 0;
```
redundant initialization, drop

fabiobaltieri commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -78,12 +78,12 @@ static void decrement_coil_charge(const struct device *dev)
 }
 
-static int power_down_coils(const struct device *dev)
+static int energize_coils(const struct device *dev, const bool energize)
```
should probably be `const bool energized`

fabiobaltieri commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -327,23 +327,39 @@ static int gpio_stepper_set_event_callback(const struct device *dev,
 
-	K_SPINLOCK(&data->lock) {
+	if (data->is_enabled == enable) {
+		return err;
```
`return 0`, then drop the initializer above

fabiobaltieri commented on drivers/stepper/gpio_stepper_controller.c:
```diff
@@ -327,23 +327,39 @@ static int gpio_stepper_set_event_callback(const struct device *dev,
+static int gpio_stepper_stop(const struct device *dev)
+{
+	struct gpio_stepper_data *data = dev->data;
+	int err = 0;
```
drop the initializer

fabiobaltieri commented on drivers/stepper/stepper_shell.c:
```diff
@@ -195,6 +198,30 @@ static int cmd_stepper_enable(const struct shell *sh, size_t argc, char **argv)
+
+	err = stepper_stop(dev);
+	if (err) {
+		shell_error(sh, "Error: %d", err);
```
do you really want to continue?

fabiobaltieri commented on drivers/stepper/stepper_shell.c:
```diff
@@ -195,6 +198,30 @@ static int cmd_stepper_enable(const struct shell *sh, size_t argc, char **argv)
+		shell_error(sh, "Error: %d", err);
+	}
+
+
```
drop one blank line, one will do
